### ISSUE TITLE
Race condition

### ISSUE THREAD

**Issue by RulerOf** [elpy1/ssh-over-ssm#16]
I used this script around a year ago to build a project that needed some casual SSH access and set it up so we could run commands over SSH during CI runs. We'd receive random failures attempting to SSH and eventually determined the cause.

https://github.com/elpy1/ssh-over-ssm/blob/057ee41d266b25822ce393212936e4218ca9b6e8/ssh-ssm.sh#L19-L26

There's no guarantee that the call to `aws ssm send-command` will start and complete the installation of the temporary SSH key on the target machine prior to invoking `aws ssm start-session`.

We tried a few different ways to work around this, including a simple `sleep` command, but we'd still run into edge cases—SSM can take 30-45 seconds before a command will begin to execute. We eventually dropped this tool for CI runs, instead using a script that would send the command, wait for it to complete, and then dump stderr/stdout as captured by SSM.

That said, I still use this tool for some interactive logons and would like to adopt it more broadly, but I'd like to fix the race condition first.

The actual work that's done on the target node is unfortunately a little cryptic. @elpy1, can you think of a better way to remove the temporary SSH key from the node that doesn't require blocking the command that installs it? Perhaps using `at`?

**Comment by RulerOf** [comment]
Okay. `at` works.

[This is the script](https://gist.github.com/RulerOf/4ef869ac5bc6b6b1d15c0fe6db65ae24) that I spoke of earlier which we use for executing commands. I put it in my path and then modified ssh-ssm.sh to instead work like this:

```shell
  checks "$@" && ssh_pubkey=$(ssh-add -L 2>/dev/null| head -1) || mktmpkey
  ssm_cmd=$(cat <<EOF
    "u=\$(getent passwd ${2}) && x=\$(echo \$u |cut -d: -f6) || exit 1
    install -d -m700 -o${2} \${x}/.ssh; grep '${ssh_pubkey}' \${x}/${ssh_authkeys} && exit 1
    printf '${ssh_pubkey}'|tee -a \${x}/${ssh_authkeys}
    echo \"sleep 15; sed -i s,'${ssh_pubkey}',, \${x}/${ssh_authkeys}\" | at now"
EOF
  )

  # put our public key on the remote server
  ec2-exec.sh "$1" "${ssm_cmd}"

  # start ssh session over ssm
  aws ssm start-session --document-name AWS-StartSSHSession --target "$1" #--debug
```

This does continue to work and should resolve the race condition. The downside, of course, is slower connection to the instance. It's reasonable, but still longer than before, coming in at about 8s til the session is ready:

```
% time AWS_PROFILE=poc ssh root@i-0c8d3cd2example whoami                                                                                        
# Command on i-0c8d3cd2example is Success
root
AWS_PROFILE=poc ssh root@i-0c8d3cd2example whoami  0.01s user 0.01s system 0% cpu 8.610 total
```

**Comment by elpy1** [comment]
Hey @RulerOf, thanks for all the info.

I only just saw your most recent message above. I've pushed a change to remove the blocking on ssh key removal - hopefully this fixes the delay while waiting for the remote command to run successfully.

I did a bit of testing by trying to make use of `aws ssm wait command-executed --instance-id <id> --command-id <id>` before `ssm start-session` but it's sort of redundant because the command executes and completes nearly instantly.

If you're finding that `ssm start-session` runs too quickly after `ssm send-command`, the easiest option is probably to add a 1s `sleep` before it.

Let me know how it goes.

**Comment by RulerOf** [comment]
> I did a bit of testing by trying to make use of `aws ssm wait command-executed --instance-id <id> --command-id <id>` before `ssm start-session` but it's sort of redundant because the command executes and completes nearly instantly.

Non-blocking key installation with  `aws ssm wait` is definitely the right way to do this. I strongly recommend adding it to the script to fully resolve this issue. 

When we did our testing, and as I'm sure you discovered, it was *extremely* difficult to reproduce this bug: the send command is usually too fast to show this problem. It would happen often enough to break automation but never enough to observe it interactively.

**Comment by elpy1** [comment]
> Non-blocking key installation with `aws ssm wait` is definitely the right way to do this. I strongly recommend adding it to the script to fully resolve this issue.

Jumping off for now but will look into it in more detail soon, hopefully tomorrow.

> When we did our testing, and as I'm sure you discovered, it was _extremely_ difficult to reproduce this bug: the send command is usually too fast to show this problem. It would happen often enough to break automation but never enough to observe it interactively.

Yep I understand the issue you're describing now, would be annoying haha.

**Comment by elpy1** [comment]
OK so it turns out the change I pushed yesterday wasn't entirely working as expected. I've made a few changes to improve general behaviour. The ones relevant to this issue:

- don't fail the `ssm send-command` if public key already exists on remote. This fixes any failures due to disconnecting and re-connecting or connecting multiple times to a server before key removal (within 15s after first connection).
- fix blocking in `ssm send-command` by ensuring key removal goes to the background and doesn't write to stdout/stderr
- use waiter to verify successful ssm command execution before starting session

FYI the reason I didn't use `at` is because it's not available by default on most distros.

Let me know how it goes and whether it fixes your issue.
